Virtual Network Gateway NAT feature changes

Khushboo-Baheti opened the pull request:
<!-- DO NOT DELETE THIS TEMPLATE -->

## Description

In Virtual Network Gateway deployments a customer might have hundreds of connections. Due to various reasons such as merger & acquisition or reseller hosting, multiple of these connections might have duplicate IP Address ranges configured. There are also scenarios where customers prefer to hide branch address range from other branches. These customer scenarios can be addressed with Source NAT transforms at any of the On-premise that connects to a virtual network Gateway. For NAT’ing the On-premise branches (Locla Network Gateways), the customer may apply NAT config at the Connection level and for NAT’ing the VNET ranges the customer may apply the NAT config at the Virtual Network Gateway level.

## Checklist

- [X] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [X] The title of the PR is clear and informative
- [X] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [X] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [ ] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [X] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
        - https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/884
    - [X] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

BethanyZhou commented:
Hi @Khushboo-Baheti , our CI failed due to following build errors.
```
D:\a\1\s\src\Network\Network\VirtualNetworkGateway\VirtualNetworkGatewayNatRule\VirtualNetworkGatewayNatRuleBaseCmdlet.cs(38,106): error CS0234: The type or namespace name 'VirtualNetworkGatewayNatRule' does not exist in the namespace 'Microsoft.Azure.Management.Network.Models' (are you missing an assembly reference?) [D:\a\1\s\src\Network\Network\Network.csproj] [D:\a\1\s\build.proj]
D:\a\1\s\src\Network\Network\VirtualNetworkGateway\VirtualNetworkGatewayNatRule\VirtualNetworkGatewayNatRuleBaseCmdlet.cs(30,16): error CS0246: The type or namespace name 'IVirtualNetworkGatewayNatRulesOperations' could not be found (are you missing a using directive or an assembly reference?) [D:\a\1\s\src\Network\Network\Network.csproj] [D:\a\1\s\build.proj]
```
I suspect upgrading sdk to newest one will fix these errors, could you have a look?

Khushboo-Baheti commented:
> Hi @Khushboo-Baheti , our CI failed due to following build errors.
> 
> ```
> D:\a\1\s\src\Network\Network\VirtualNetworkGateway\VirtualNetworkGatewayNatRule\VirtualNetworkGatewayNatRuleBaseCmdlet.cs(38,106): error CS0234: The type or namespace name 'VirtualNetworkGatewayNatRule' does not exist in the namespace 'Microsoft.Azure.Management.Network.Models' (are you missing an assembly reference?) [D:\a\1\s\src\Network\Network\Network.csproj] [D:\a\1\s\build.proj]
> D:\a\1\s\src\Network\Network\VirtualNetworkGateway\VirtualNetworkGatewayNatRule\VirtualNetworkGatewayNatRuleBaseCmdlet.cs(30,16): error CS0246: The type or namespace name 'IVirtualNetworkGatewayNatRulesOperations' could not be found (are you missing a using directive or an assembly reference?) [D:\a\1\s\src\Network\Network\Network.csproj] [D:\a\1\s\build.proj]
> ```
> 
> I suspect upgrading sdk to newest one will fix these errors, could you have a look?

@BethanyZhou Yes upgrading SDK will fix this I used unsigned SDK build for local testing - https://dev.azure.com/nrp-swagger/swagger-stack/_build/results?buildId=272&view=results My SDK changes - https://github.com/Azure/azure-rest-api-specs/pull/13481/files
 I am not sure how to update published SDK for Network-February-Release

BethanyZhou commented:
@Satya-anshu is uploading unsigned SDK as local feed. I believe merging network-february into your branch could help you case if SDK is uploaded.

Khushboo-Baheti commented:
Thanks a lot @BethanyZhou for guiding @Satya-anshu SDK doesn't have my changes. I have updated local-feed with the latest SDK. I have updated Network.CSproj and NetworkTest.CSProj Do I need to update any other files ?

BethanyZhou commented:
@Khushboo-Baheti , Sure, please update all files referred to this SDK.

Khushboo-Baheti commented:
@BethanyZhou I have updated PRs with all the required changes. Can you please look into this. Thanks

BethanyZhou commented:
Hi @Khushboo-Baheti , please suppress following issues in /tools/StaticAnalysis/Exceptions/Az.Network/SignatureIssues.csv
```
"D:\a\1\s\artifacts\Debug\Az.Network\Microsoft.Azure.PowerShell.Cmdlets.Network.dll","Microsoft.Azure.Commands.Network.VirtualNetworkGateway.NewAzureVirtualNetworkGatewayNatRule","New-AzVirtualNetworkGatewayNatRule","1","8100","New-AzVirtualNetworkGatewayNatRule Does not support ShouldProcess but the cmdlet verb New indicates that it should.","Determine if the cmdlet should implement ShouldProcess and if so determine if it should implement Force / ShouldContinue"
"D:\a\1\s\artifacts\Debug\Az.Network\Microsoft.Azure.PowerShell.Cmdlets.Network.dll","Microsoft.Azure.Commands.Network.VirtualNetworkGateway.UpdateAzureVirtualNetworkGatewayNatRule","Update-AzVirtualNetworkGatewayNatRule","1","8100","Update-AzVirtualNetworkGatewayNatRule Does not support ShouldProcess but the cmdlet verb Update indicates that it should.","Determine if the cmdlet should implement ShouldProcess and if so determine if it should implement Force / ShouldContinue"
```

BethanyZhou reviewed (changes requested):
There needs further discussion about EnableBgpRouteTranslationForNatFlag and DisableBgpRouteTranslationForNatFlag, please go back to https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/884 and take a look

Khushboo-Baheti commented:
@BethanyZhou 2 of the checks are failing with following error and I am not sure how to fix this. Can you please help - 2021-04-12T20:45:44.9752026Z   Segmentation fault (core dumped)
2021-04-12T20:45:44.9847054Z /home/vsts/work/1/s/build.proj(182,5): error MSB3073: The command "dotnet sln /home/vsts/work/1/s/artifacts/Azure.PowerShell.sln add "/home/vsts/work/1/s/src/RecoveryServices/RecoveryServices/RecoveryServices.csproj"" exited with code 139.
2021-04-12T20:45:45.0181779Z ##[error]Error: The process '/usr/bin/dotnet' failed with exit code 1
2021-04-12T20:45:45.0204032Z Info: Azure Pipelines hosted agents have been updated and now contain .Net 5.x SDK/Runtime along with the older .Net Core version which are currently lts. Unless you have locked down a SDK version for your project(s), 5.x SDK might be picked up which might have breaking behavior as compared to previous versions. You can learn more about the breaking changes here: https://docs.microsoft.com/en-us/dotnet/core/tools/ and https://docs.microsoft.com/en-us/dotnet/core/compatibility/ . To learn about more such changes and troubleshoot, refer here: https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/build/dotnet-core-cli?view=azure-devops#troubleshooting

Khushboo-Baheti commented:
> There needs further discussion about EnableBgpRouteTranslationForNatFlag and DisableBgpRouteTranslationForNatFlag, please go back to [Azure/azure-powershell-cmdlet-review-pr#884](https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/884) and take a look

I have updated issue and also updated this PR as per your suggestions. Thanks

BethanyZhou commented on src/Network/Network/VirtualNetworkGateway/UpdateAzureVirtualNetworkGatewayCommand.cs:
```diff
@@ -201,6 +201,17 @@ public class SetAzureVirtualNetworkGatewayCommand : VirtualNetworkGatewayBaseCmd
+
+        [Parameter(
+            Mandatory = false,
+            HelpMessage = "Flag to enable and disable BgpRouteTranslationForNat on this VirtualNetworkGateway.")]
```
I believe its help message should be updated also.

Khushboo-Baheti replied:
@BethanyZhou I have updated help message as per your suggestion. Thanks

BethanyZhou commented:
> @BethanyZhou Beisi Zhou FTE 2 of the checks are failing with following error and I am not sure how to fix this. Can you please help - 2021-04-12T20:45:44.9752026Z Segmentation fault (core dumped)
> 2021-04-12T20:45:44.9847054Z /home/vsts/work/1/s/build.proj(182,5): error MSB3073: The command "dotnet sln /home/vsts/work/1/s/artifacts/Azure.PowerShell.sln add "/home/vsts/work/1/s/src/RecoveryServices/RecoveryServices/RecoveryServices.csproj"" exited with code 139.
> 2021-04-12T20:45:45.0181779Z ##[error]Error: The process '/usr/bin/dotnet' failed with exit code 1
> 2021-04-12T20:45:45.0204032Z Info: Azure Pipelines hosted agents have been updated and now contain .Net 5.x SDK/Runtime along with the older .Net Core version which are currently lts. Unless you have locked down a SDK version for your project(s), 5.x SDK might be picked up which might have breaking behavior as compared to previous versions. You can learn more about the breaking changes here: https://docs.microsoft.com/en-us/dotnet/core/tools/ and https://docs.microsoft.com/en-us/dotnet/core/compatibility/ . To learn about more such changes and troubleshoot, refer here: https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/build/dotnet-core-cli?view=azure-devops#troubleshooting

Forget about `Segmentation fault`. This is a known and random issue about the environment of our CI pipelines. Just use `/azp run` to restart all pipelines.

BethanyZhou reviewed (approved):
LGTM

Khushboo-Baheti commented:
Thanks @BethanyZhou for approval. Please let me know if anything is required by my side. Do I need to ping all reviewers and take approval from them ?

BethanyZhou commented:
@Khushboo-Baheti Will merge

erich-wang commented:
Force merge due of known CI issue.
